libs/azure-storage/tests/unit_tests/test_document_loaders.py

```diff
@@ -0,0 +1,162 @@
+from typing import Any, Callable, Iterator, Tuple
+from unittest.mock import MagicMock, patch
+
+import azure.identity
+import pytest
+from azure.storage.blob import BlobClient, ContainerClient
+from azure.storage.blob._download import StorageStreamDownloader
+
+from langchain_azure_storage.document_loaders import AzureBlobStorageLoader
+
+
+@pytest.fixture
+def account_url() -> str:
+    return "https://testaccount.blob.core.windows.net"
+
+
+@pytest.fixture
+def container_name() -> str:
+    return "test-container"
+
+
+@pytest.fixture
+def blobs() -> list[dict[str, str]]:
+    return [
+        {"blob_name": "text_file.txt", "blob_content": "test content"},
+        {"blob_name": "json_file.json", "blob_content": "{'test': 'test content'}"},
+        {"blob_name": "csv_file.csv", "blob_content": "col1,col2\nval1,val2"},
+    ]
+
+
+@pytest.fixture
+def create_azure_blob_storage_loader(
+    account_url: str, container_name: str
+) -> Callable[..., AzureBlobStorageLoader]:
+    def _create_azure_blob_storage_loader(**kwargs: Any) -> AzureBlobStorageLoader:
+        return AzureBlobStorageLoader(
+            account_url,
+            container_name,
+            **kwargs,
+        )
+
+    return _create_azure_blob_storage_loader
+
+
+@pytest.fixture
+def mock_blob_clients(
+    account_url: str, container_name: str, blobs: list[dict[str, str]]
+) -> Callable[[str], MagicMock]:
+    def _get_blob_client(blob_name: str) -> MagicMock:
+        mock_blob_client = MagicMock(spec=BlobClient)
+        mock_blob_client.url = f"{account_url}/{container_name}/{blob_name}"
+        mock_blob_data = MagicMock(spec=StorageStreamDownloader)
+        content = next(
+            blob["blob_content"] for blob in blobs if blob["blob_name"] == blob_name
+        )
+        mock_blob_data.readall.return_value = content.encode("utf-8")
+        mock_blob_client.download_blob.return_value = mock_blob_data
+        return mock_blob_client
+
+    return _get_blob_client
+
+
+@pytest.fixture(autouse=True)
+def mock_container_client(
+    blobs: list[dict[str, str]], mock_blob_clients: Callable[[str], MagicMock]
+) -> Iterator[Tuple[MagicMock, MagicMock]]:
+    with patch(
+        "langchain_azure_storage.document_loaders.ContainerClient"
+    ) as mock_container_client_cls:
+        mock_client = MagicMock(spec=ContainerClient)
+        mock_client.list_blob_names.return_value = [blob["blob_name"] for blob in blobs]
+        mock_client.get_blob_client.side_effect = mock_blob_clients
+        mock_container_client_cls.return_value = mock_client
+        yield mock_container_client_cls, mock_client
+
+
+def test_lazy_load(
+    account_url: str,
+    container_name: str,
+    blobs: list[dict[str, str]],
+    create_azure_blob_storage_loader: Callable[..., AzureBlobStorageLoader],
+) -> None:
+    loader = create_azure_blob_storage_loader()
+    docs = loader.lazy_load()
+    assert len(list(docs)) == len(blobs)
```

Comment:
Instead of asserting the length before we iterate, it probably makes sense to make the assertion after the for loop and assert against the final value of the `index`? Mainly right now by calling the `list()` ahead of time, we exhaust the iterator and when we get to the for loop, none of the assertions are happening because there is nothing to iterate through.

Another idea we could do is maybe build up an expected document list and then assert the list returned from `lazy_load()` matches that expected list? This will only work if the `__eq__` is appropriately implemented on the `Document` base class. I think it is worth try out to see if it cleans some of this up.